File: 3-29/46全排列/46.py
#  -*- coding: utf-8 -*-

#  Editor      : Pycharm
#  File        : 46
#  Author      : ZhenyuMa
#  Created     : 2020/3/30 12:00 上午
#  Description : 回溯


from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        self.res = []
        self.visited = [False] * len(nums)
        self.helper(nums, 0, [])
        return self.res

    def helper(self, nums, idx, cur_res):
        if idx == len(nums):
            self.res.append(cur_res[:])
            return

        for i in range(len(nums)):
            if not self.visited[i]:
                cur_res.append(nums[i])
                self.visited[i] = True
                self.helper(nums, idx + 1, cur_res)
                cur_res.pop()
                self.visited[i] = False


if __name__ == "__main__":
    test = Solution()
    print(test.permute([1, 2, 3]))

File: 2020-04/4-25/46全排列/46.cpp


class Solution {
private:
	vector<vector<int>> res;
public:
    vector<vector<int>> permute(vector<int>& nums) {
    	vector<bool> visited(nums.size(), false);
    	vector<int> cur_res;
    	trace_back(nums, visited, cur_res);
    	return res;
    }

    void trace_back(const vector<int>& nums, vector<bool>& visited, vector<int>& cur_res){
    	if(cur_res.size() == nums.size()){
    		res.push_back(cur_res);  // 装的就是copy过的，所以不用复制哦
    		return;
    	}

    	for(int i = 0; i < nums.size(); i++){
    		if(!visited[i]){
    			visited[i] = true;
    			cur_res.push_back(nums[i]);
    			trace_back(nums, visited, cur_res);
    			visited[i] = false;
    			cur_res.pop_back();
    		}
    	}
    }
};